xdsl/dialects/fsm.py

```diff
@@ -0,0 +1,379 @@
+                ):
+                    raise VerifyException(
+                        "Output types must be consistent with the machine's"
+                    )
+            parent = parent.parent_op()
+
+
+@irdl_op_definition
+class StateOp(IRDLOperation):
+    """
+    Represents a state of a state machine. This op includes an
+    `$output` region with an `fsm.output` as terminator to define the machine
+    outputs under this state. This op also includes a `transitions` region that
+    contains all the transitions of this state
+    """
+
+    name = "fsm.state"
+
+    output = region_def()
+
+    transitions = region_def()
+
+    sym_name = attr_def(StringAttr)
+
+    traits = frozenset([NoTerminator(), SymbolOpInterface()])
+
+    def __init__(
+        self,
+        sym_name: str,
+        output: Region | type[Region.DEFAULT] = Region.DEFAULT,
+        transitions: Region | type[Region.DEFAULT] = Region.DEFAULT,
+    ):
+        attributes: dict[str, Attribute] = {
+            "sym_name": StringAttr(sym_name),
+        }
+        if not isinstance(output, Region):
+            output = Region(Block())
+        if not isinstance(transitions, Region):
+            transitions = Region(Block())
+        super().__init__(
+            attributes=attributes,
+            regions=[output, transitions],
+        )
+
+    def verify_(self):
+        parent = self.parent_op()
+
+        while parent is not None:
+            if (
+                isinstance(parent, MachineOp)
+                and getattr(parent, "res_attrs") is not None
+                and len(getattr(parent, "res_attrs")) > 0
+                and self.output.block.first_op is None
+            ):
+                raise VerifyException(
+                    "State must have a non-empty output region when the machine has results."
+                )
+            parent = parent.parent_op()
+
+
+@irdl_op_definition
+class UpdateOp(IRDLOperation):
+    """
+    Updates the `$variable` with the `$value`. The definition op of
+    `$variable` should be an `fsm.variable`. This op should *only* appear in the
+    `action` region of a transtion
+    """
+
+    name = "fsm.update"
+
+    variable = operand_def(Attribute)
+
+    value = operand_def(Attribute)
+
+    traits = frozenset([HasParent(TransitionOp)])
+
+    def __init__(
+        self,
+        variable: SSAValue | Operation,
+        value: SSAValue | Operation,
+    ):
+        super().__init__(
+            operands=[variable, value],
+        )
+
+    def verify_(self) -> None:
+        if not isinstance(self.variable.owner, VariableOp):
+            raise VerifyException("Destination is not a variable operation")
+
+        parent = self.parent_op()
+        while parent is not None:
+            if isinstance(parent, TransitionOp):
+                # walk through the action region
+                found = 0
+                for op in parent.action.walk():
```

Comment:
```suggestion
                for op in parent.action.ops:
```
So it doesn't recursively go into regions (since there are none)

Reply:
For now I had to leave this one like that because it would raise a "single-block" related exception with the ops operator. maybe we can take a closer look later!

Reply:
This happens because the `action` region of the `TransitionOp` is not a SingleBlockRegion. I am not sure about how FSM works internally, but it feels like a lot of these regions might actually be SingleBlockRegions, which would be very useful to specify. If this is not done in the circt dialect, I would not do it here either though :)

Reply:
Thanks, I'll double check that :)